src/network/networkd-dhcp4.c

```diff
@@ -1319,9 +1319,30 @@ int dhcp4_configure(Link *link) {
         if (r < 0)
                 return log_link_warning_errno(link, r, "DHCP4 CLIENT: Failed to set callback: %m");
 
-        r = sd_dhcp_client_set_request_broadcast(link->dhcp_client, link->network->dhcp_broadcast);
-        if (r < 0)
-                return log_link_warning_errno(link, r, "DHCP4 CLIENT: Failed to set request flag for broadcast: %m");
+        /* Some interfaces require that the DHCPOFFER message is being broadcast because they can't
+         * handle unicast messages while not fully configured. Setting device property ID_NET_DHCP_BROADCAST
+         * to "1" can be used to conditionally enable the dhcp_broadcast flag for such interfaces. */
+        dhcp_broadcast = link->network->dhcp_broadcast;
+        if (dhcp_broadcast < 0 && link->sd_device) {
+                const char *val;
+
+                if (sd_device_get_property_value(link->sd_device, "ID_NET_DHCP_BROADCAST", &val) >= 0) {
+                        r = parse_boolean(val);
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "DHCP4 CLIENT: Failed to parse ID_NET_DHCP_BROADCAST, ignoring: %m");
+                        else {
+                                dhcp_broadcast = r;
+                                log_link_debug(link, "DHCP4 CLIENT: %s broadcast flag based on ID_NET_DHCP_BROADCAST udev property.",
+                                               dhcp_broadcast ? "Enabling" : "Disabling");
+                        }
+                }
+        }
```

Comment:
can we split this out into a helper function maybe?

i.e. link_needs_dhcp_broadcast() or so as function, that gets the Link object and returns < 0 on error, and otherwise the combined value from the network config plus the udev data.

Reply:
re udev rule: qeth devices can be configured to run either in l2 mode or l3 mode. The qeth driver uses sub-ordinate modules qeth_l2 or qeth_l3 ('disciplines') to talk to the card in the respective mode. From a user perspective the mode is switched by echoing 0 or 1 to $device/layer2. The device must be disabled to change the mode.

Reply:
re helper function: I can do that. I would however like to see consensus on the udev rule. As I can see that the ID_NET_DRIVER discussion has added to confusion, I find the layer2 attribute matching to be more along the lines of what the user sees in terms of sysfs entities.

Reply:
@poettering as i have mentioned before there are no modaliases. Hence hwdb approach cannot be used. The same ccw id devices look the same, and one can bind the same ccw qeth device as either qeth_l2 or qeth_l3 driver, depending on the underlying mainframe config.

@mihajlov i still don't like having three comparisons in the udev rule. But i'm ok with merging this as is.

Reply:
hmm, so is there a package that is typically installed on s390 that deals with s390 stuff? i.e. s390-utils or something like that?

might be an idea to ship the check for the udev property in systemd/network upstream, but the udev rule for it in such an s390-util package if there is one that everyone running s390 has installed anyway.

Reply:
yes s390-tools is installed everywhere on s390x. and yes udev rule could be shipped there, instead of here. Or like chzdev could generate it when it generates udev rules for qeth. Will need to ask @mihajlov about that.

Reply:
@oberpar , @julianwiedmann see discussion above. The general idea is to ship a udev rule in s390-tools which emits property ID_NET_DHCP_BROADCAST if a qeth interface requires the DHCP broadcast flag (== interface operates in l3 mode) which can be used by systemd-networkd to configure the DHCP client accordingly. Thoughts?